### PR TITLE
Pass entire Reader mode templates through post-processor

### LINKED ISSUE #688: CSS in Reader mode is not minified
The AMP CSS inserting by the plugin (`.alignright` through `.back-to-top`) is not minified.
screen shot 2017-04-24 at 4 48 15 pm

### LINKED ISSUE #2202: Send entire templates in Reader/Classic mode through the sanitizers
Currently only content (and the featured image) are sent through the sanitizers. This is the classic approach, and it is why there's the _content_ in `amp_content_sanitizers`. However, limiting the sanitizers to the content has some problems which would be fixed by sanitizing the entire page: 

1. Markup that is output outside the content (such as in `amp_post_template_head` or `amp_post_template_footer`) is not sent through the sanitizers, resulting in the possibility of invalid AMP pages when plugins don't implement AMP properly. This has come up frequently in the support forums. By sanitizing the entire document, this would no longer happen frequently.
2. The CSS in the Reader mode templates is output in a `style.php` template and the `amp_post_template_css` action. The style rules are not sent through the CSS sanitizer. This means that they are neither minified nor tree-shaken. If the entire template were sent through the sanitizers, then this would be a key benefit (and would resolve #688).
3. By sending the entire reader mode templates through the sanitizers, this closer aligns with traditional theme templates and will facilitate transitioning from those classic post templates to standard theme templates (e.g. where `$this->load_parts()` is not used).
4. There are currently two entirely different code-paths for Reader vs non-Reader modes. There are opportunities to better harmonize the two, if we can use the code in `AMP_Theme_Support` for both. 

----

This should make Reader mode templates available for Polylang to translate ([topic](wordpress.org/support/topic/static-home-page-with-polylang-2/)). 

By making Reader mode templates regular WordPress templates, we'll also be able to avoid situations where users copy Reader mode templates to be used in Transitional mode, only to find that they start getting fatal errors due to `$this`. (Examples: [1](https://wordpress.org/support/topic/cant-customize-templates-uncaught-error-using-this-when-not-in-object-conte/), [2](https://wordpress.org/support/topic/childlike-plugin-mods/), [3](https://wordpress.org/support/topic/page-unable-to-reach/)). 

**Update:** This is probably a bit out of scope here. I think we should take the existing Reader mode post templates and port them to regular WordPress theme templates, but when we do so we need to introduce generic `amp_head` and `amp_footer` actions (and the like) which run instead of `amp_post_template_head` (à la `wp_head`) and `amp_post_template_footer` (à la `wp_footer`), similar to what we did for stories with `amp_story_head`. The `amp_post_template_*` actions should be deprecated because they are passed a `AMP_Post_Template` instance which is not relevant when templates can render any WordPress query, including index pages.

### PR DESCRIPTION
**Pull request by westonruter**
## Summary

- [x] Eliminate majority of Reader mode template's unique rendering path, rendering them with the same processing logic in `AMP_Theme_Support`. This ensures the entire documents are sanitized and all CSS is tree-shaken.
- [x] Use `template_include` filter to load a template which initializes the `AMP_Post_Template` class while output buffering is being done. This ensures themes which extend the Reader mode templates and make use of `$this` may continue to do so.
- [x] Deprecate functions and `AMP_Content` class which are no longer used.
- [x] Ensure validation error reporting (compatibility tool) remains disabled in Reader mode (for now).
- [x] Move deprecated functions to `deprecated.php`.
- [x] Fix viewing Reader template in Customizer.

Fixes #2202.
Fixes #688.

## Checklist

- [ ] My pull request is addressing an [open issue](https://github.com/ampproject/amp-wp/contributing/project-management.md#life-of-an-issue) (please create one otherwise).
- [ ] My code is tested and passes existing [tests](https://github.com/ampproject/amp-wp/contributing/engineering.md#tests).
- [ ] My code follows the [Engineering Guidelines](https://github.com/ampproject/amp-wp/contributing/engineering.md) (updates are often made to the guidelines, check it out periodically).

### CODE REVIEW

File: includes/templates/class-amp-post-template.php [hunk 1]
```diff
@@ -164,10 +148,41 @@ public function __construct( $post ) {
...
+			case 'post_modified_timestamp':
+				return get_post_modified_time( 'U', false, $this->post );
+			case 'post':
+				return $this->post;
+			case 'post_id':
+				return $this->post->ID;
+			case 'site_icon_url':
+				/** This filter is documented in includes/amp-helper-functions.php */
+				return apply_filters( 'amp_site_icon_url', function_exists( 'get_site_icon_url' ) ? get_site_icon_url( self::SITE_ICON_SIZE ) : '' );
+		}
```
Comment (westonruter) [inline includes/templates/class-amp-post-template.php]: This makes me nervous. By making the fields here lazy-loaded when invoking the getter, it bypasses the `amp_post_template_data` filter. Perhaps this should be reverted, and the introduction of regular WordPress template tags be replaced with `AMP_Post_Template::get()` to ensure no breakage of backwards-compatibility. See search for `amp_post_template_data` filter across plugins:  https://wpdirectory.net/search/01DSTQ6DK2GCDKF9QDEN0E8X7R

Nevertheless, going forward sites would be free to use WordPress template tags when writing Reader mode templates.

Comment (westonruter) [reply includes/templates/class-amp-post-template.php]: Undone in aed6c0e6550ac9404fa96b838b6ab1ddd541a933

File: includes/class-amp-theme-support.php [hunk 2]
```diff
@@ -417,15 +419,26 @@ public static function finish_init() {
...
 
-		$theme_support = self::get_theme_support_args();
+		$is_reader_mode = self::READER_MODE_SLUG === self::get_support_mode();
+		$theme_support  = self::get_theme_support_args();
 		if ( ! empty( $theme_support['template_dir'] ) ) {
 			self::add_amp_template_filters();
+		} elseif ( $is_reader_mode ) {
+			add_action(
+				'template_include',
+				function() {
```
Comment (westonruter) [inline includes/class-amp-theme-support.php]: Should either be `static` or to add a new global function.

File: amp.php [hunk 3]
```diff
@@ -449,6 +446,7 @@ function amp_force_query_var_value( $query_vars ) {
  * @return void
  */
 function amp_maybe_add_actions() {
+	_deprecated_function( __FUNCTION__, '1.5' );
```
Comment (westonruter) [inline amp.php]: The deprecated functions here should be moved to `includes/deprecated.php` (which interestingly is not being included anywhere!)

Also, it would be good to move all of the non-bootstrapping functions in `amp.php` into `includes/amp-helper-functions.php`. This will allow us to use PHP>5.2 in these functions.

**Comment by westonruter** [comment]
The `AMP_Content` class can be hard-deprecated once we have the new API established in #3662.

File: includes/class-amp-theme-support.php [hunk 4]
```diff
@@ -417,15 +419,26 @@ public static function finish_init() {
 
 		self::ensure_proper_amp_location();
 
-		$theme_support = self::get_theme_support_args();
+		$is_reader_mode = self::READER_MODE_SLUG === self::get_support_mode();
+		$theme_support  = self::get_theme_support_args();
 		if ( ! empty( $theme_support['template_dir'] ) ) {
 			self::add_amp_template_filters();
+		} elseif ( $is_reader_mode ) {
+			add_action(
```
Comment (westonruter) [inline includes/class-amp-theme-support.php]:
```suggestion
			add_filter(
```

File: includes/class-amp-theme-support.php [hunk 5]
```diff
@@ -417,15 +419,26 @@ public static function finish_init() {
 
 		self::ensure_proper_amp_location();
 
-		$theme_support = self::get_theme_support_args();
+		$is_reader_mode = self::READER_MODE_SLUG === self::get_support_mode();
+		$theme_support  = self::get_theme_support_args();
 		if ( ! empty( $theme_support['template_dir'] ) ) {
 			self::add_amp_template_filters();
+		} elseif ( $is_reader_mode ) {
```
Comment (westonruter) [inline includes/class-amp-theme-support.php]: Instead of having a separate condition for reader mode, it may be good to combine with `add_amp_template_filters`. The thinking is we may want to go ahead and try to load a template as if it is in Transitional mode, but then if we are about to load a `single.php` or `page.php` that contains `$this`, then we can at that point load it via `AMP_Post_Template`. This will avoid necessarily doing the current hard-coding of `single.php` and `page.php` as seen in:

https://github.com/ampproject/amp-wp/blob/451dd22cb6f80a631c95bec7fef1b0e02f8e9c48/includes/templates/class-amp-post-template.php#L198-L199

This would allow Reader templates to make use of the full template hierarchy, and ease transition to _Transitional_ templates.

In other words, if someone does:

```php
add_theme_support( 'amp', [ 'template_dir' => 'amp' ] );
```

This should either allow the templates in the `amp` directory to be classic `AMP_Post_Template`-based templates, or regular WordPress theme templates. We can start phasing out the use of `AMP_Post_Template`, only loading it when we detect it is needed.

In fact, there shouldn't be any real difference between Reader mode and Transitional mode when a `template_dir` is present.

File: includes/templates/class-amp-post-template.php [hunk 6]
```diff
@@ -111,23 +111,24 @@ public function __construct( $post ) {
 		$this->data = [
```
Comment (westonruter) [inline includes/templates/class-amp-post-template.php]: Instead of assembling all of this data in the constructor, I think it would be better to lazily construct it upon invoking the getter the first time. This would allow us to construct the `AMP_Post_Template` preemptively without potentially it ever being used (and thus avoiding wasted queries), such as when normal WordPress templates are being used in the `amp` theme directory.

We could then still pass the `AMP_Post_Template` object in the existing hooks like `amp_post_template_head`, and as long as no hooks call the getter, then the data is not fetched.

File: amp.php [hunk 7]
```diff
@@ -667,6 +668,7 @@ function amp_render() {
  * @global WP_Query $wp_query
  */
 function amp_render_post( $post ) {
+	_deprecated_function( __FUNCTION__, '1.5' );
```
Comment (westonruter) [inline amp.php]: We need to check if these functions are being used in themes/plugins, and if so, potentially defer this hard deprecation until we have the AMP middleware rendering stack available to use instead (same for `AMP_Content`).

File: includes/class-amp-theme-support.php (hunk 5 above)
Comment (schlessera) [reply includes/class-amp-theme-support.php]: Instead of detecting `$this` inside of a template (which requires parsing the entire template file), we could also look into wrapping _all_ templates in a compatibility wrapper object by default, as regular WordPress templates don't use `$this`. However, I'm not 100% sure whether that cleanly works across global state (which is expected in regular WordPress templates).

File: amp.php (hunk 3 above)
Comment (schlessera) [reply amp.php]: > which interestingly is not being included anywhere!

Hehe, does that mean that the current contents of that file can be hard-deprecated ?

File: templates/html-end.php [hunk 8]
```diff
@@ -19,7 +19,17 @@
  */
 ?>
 
-<?php do_action( 'amp_post_template_footer', $this ); ?>
+<?php
+/**
+ * Fires when just before after printing the </body>.
```
Comment (schlessera) [inline templates/html-end.php]: Huh? 🤔

File: templates/html-start.php [hunk 9]
```diff
@@ -23,12 +23,44 @@
...
+		do_action( 'amp_post_template_css', $this );
+		?>
 	</style>
 </head>
 
 <body class="<?php echo esc_attr( $this->get( 'body_class' ) ); ?>">
-<?php do_action( 'amp_post_template_body_open', $this ); ?>
+<?php
+/**
+ * Fires when immediately after printing the <body>.
```
Comment (schlessera) [inline templates/html-start.php]:
```suggestion
 * Fires immediately after printing the <body>.
```

File: includes/class-amp-theme-support.php (hunk 4 above)
Comment (schlessera) [reply includes/class-amp-theme-support.php]: If you turn the action into a filter, you'll need to actually use the return value as well.

File: includes/class-amp-theme-support.php [hunk 10]
```diff
@@ -458,6 +471,18 @@ public static function ensure_proper_amp_location( $exit = true ) {
...
+			$path_args = [];
+			wp_parse_str( $wp->matched_query, $path_args );
+			if ( isset( $path_args[ amp_get_slug() ] ) && '' !== $path_args[ amp_get_slug() ] ) {
+				wp_safe_redirect( amp_get_permalink( get_queried_object_id() ), 301 );
+				if ( $exit ) {
+					exit;
+				}
+				return true;
+			}
 		} else {
```
Comment (schlessera) [inline includes/class-amp-theme-support.php]: There's a bogus if below here that can be included in the else.
```suggestion
		} elseif ( $has_query_var && ! $has_url_param ) {
```

File: includes/templates/class-amp-content.php [hunk 11]
```diff
@@ -7,6 +7,8 @@
 
 /**
  * Class AMP_Content
+ *
+ * @deprecated
```
Comment (schlessera) [inline includes/templates/class-amp-content.php]:
```suggestion
 * @deprecated 1.5 Reader mode now sanitizes its entire template through the standard post-processor.
```

File: includes/templates/class-amp-post-template.php (hunk 6 above)
Comment (schlessera) [reply includes/templates/class-amp-post-template.php]: A "fun" way of doing that (which doesn't require conditionals all over the place) is to not define the `$data` property at all in the class and then use `__get()` to add it the first time it is being requested (through whatever means). After that, it is just present, so the `__get()` won't be hit a second time.

File: amp.php (hunk 3 above)
Comment (westonruter) [reply amp.php]: Yes, probably. The file was introduced in 9750909638bc0a9864d76e1a57ef9bd91e35f794 by yours truly but I forgot to include it. So much backwards-incompatibility breakage that didn't happen to anyone!

File: includes/class-amp-theme-support.php (hunk 4 above)
Comment (westonruter) [reply includes/class-amp-theme-support.php]: I don't understand. The filter callback is returning a value:

```php
return __DIR__ . '/templates/reader-template-loader.php';
```

It's not applying filters here.

Comment (schlessera) [reply includes/class-amp-theme-support.php]: Nevermind, I'm dumb.

File: includes/templates/class-amp-post-template.php (hunk 6 above)
Comment (westonruter) [reply includes/templates/class-amp-post-template.php]: Yes, this is what I intended to say. It's just in this class that logic would be in `\AMP_Post_Template::get()`.

File: templates/html-end.php (hunk 8 above)
Comment (westonruter) [reply templates/html-end.php]: Extraneous “after“.

Comment (schlessera) [reply templates/html-end.php]: More than that... probably extraneous "when" as well.

File: includes/class-amp-theme-support.php (hunk 10 above)
Comment (westonruter) [reply includes/class-amp-theme-support.php]: Fixed in 93cede44d26e0bdadbc3fa99db0ed6cedbfad36d

File: includes/templates/class-amp-post-template.php (hunk 6 above)
Comment (westonruter) [reply includes/templates/class-amp-post-template.php]: Done in c74ba974

File: includes/class-amp-theme-support.php (hunk 5 above)
Comment (westonruter) [reply includes/class-amp-theme-support.php]: As opposed to parsing the entire template file, I was intending to just use a simple check to see if `'$this->'` occurs inside of the file. If so, it could then load it via the `AMP_Post_Template`. Otherwise, it could do normal `include`.

File: amp.php (hunk 3 above)
Comment (westonruter) [reply amp.php]: No relevant usages: https://wpdirectory.net/search/01DT14RRMMNGT84GHZGZC900F1

File: amp.php [hunk 12]
```diff
@@ -639,6 +637,7 @@ function amp_add_post_template_actions() {
  * @deprecated This function is not used when 'amp' theme support is added.
  */
 function amp_prepare_render() {
+	_deprecated_function( __FUNCTION__, '1.5' );
```
Comment (westonruter) [inline amp.php]: No relevant usages: https://wpdirectory.net/search/01DT14VJW5JX1X1F65C7R9VY5D

File: amp.php [hunk 13]
```diff
@@ -649,6 +648,8 @@ function amp_prepare_render() {
  * @deprecated This function is not used when 'amp' theme support is added.
  */
 function amp_render() {
+	_deprecated_function( __FUNCTION__, '1.5' );
```
Comment (westonruter) [inline amp.php]: One relevant usage: https://wpdirectory.net/search/01DT14Z9DAB3X0VQWZGJPG7XA3

It's in Facebook Instant Articles & Google AMP Pages by PageFrog. However, this plugin has been closed: https://wordpress.org/plugins/pagefrog/

File: amp.php (hunk 7 above)
Comment (westonruter) [reply amp.php]: No relevant usages: https://wpdirectory.net/search/01DT152D0C97YRHGGJHYQ19X5X

File: includes/class-amp-theme-support.php (hunk 5 above)
Comment (schlessera) [reply includes/class-amp-theme-support.php]: You'll still have to read the entire file to find a potential `$this`. And this reading is separate from `include`ing the file, which you'll need to do as well, and which would normally happen via opcache and not hit the filesystem, so this adds more filesystem reads.

The compatibility wrapper I suggested would be read once and wrapped around each include. So most of the time, everything would still stay within the opcache. However, it adds complexity.

Comment (westonruter) [reply includes/class-amp-theme-support.php]: Doesn't `file_get_contents()` use caching as well? Per [docs](https://www.php.net/manual/en/function.file-get-contents.php):

> `file_get_contents()` is the preferred way to read the contents of a file into a string. It will use memory mapping techniques if supported by your OS to enhance performance.

I'm not sure what these “memory mapping techniques” entail.

In any case, we're doing `file_get_contents()` for other files with each page load, for example:

https://github.com/ampproject/amp-wp/blob/c900d8484de1751bf56fd8b1c392154fa922eebe/templates/style.php#L55

Why would this be inherently different?

In core I see `get_post_embed_html()` uses `file_get_contents()` if `SCRIPT_DEBUG` is enabled. Also, `file_get_contents()` is used by `load_script_translations()` in core.

Comment (westonruter) [reply includes/class-amp-theme-support.php]: In any case, migrating Reader mode templates to use a normal theme templates is beyond the scope of this PR.

**Comment by kienstra** [comment]
**Looks good!**

Hi @westonruter,
This looks good. It's great to have the post-processor working for the entire Reader mode templates.

For example, with:

```

add_action(
	'amp_post_template_head',
	static function() {
		?>
		<script>disallowed();</script>
		<?php
	}
);
```

This removed the script:
script-removed-as-expected

...though before this PR, there was a validation error on the front-end (in the Chrome extension).

Also, the `style[amp-custom]` looks good:

tree-shaking-done

A basic testing post looked good. It was the same with this PR, and on the `develop` branch:

testing-here

**Comment by kienstra** [comment]
The [e2e failures](https://travis-ci.org/ampproject/amp-wp/jobs/614285931#L923) don't look to be related to this PR at all. They're all for Stories.

File: includes/templates/class-amp-post-template.php [hunk 14]
```diff
@@ -321,18 +342,10 @@ private function build_post_comments_data() {
...
-		$amp_content = new AMP_Content(
-			post_password_required( $this->post ) ? get_the_password_form( $this->post ) : $this->post->post_content,
-			amp_get_content_embed_handlers( $this->post ),
-			amp_get_content_sanitizers( $this->post ),
-			[
-				'content_max_width' => $this->get( 'content_max_width' ),
-			]
-		);
+		/** This filter is documented in wp-includes/post-template.php */
+		$content = apply_filters( 'the_content', get_the_content( null, false, $this->post ) );
```
Comment (westonruter) [inline includes/templates/class-amp-post-template.php:346]: This introduced a regression. Prior to this change, posts having content with page breaks (`<!--nextpage-->`) would have their pages all displayed together. This needs to be restored, with a big reason being that there has been no multipage pagination in the Reader mode template. It's also not consistent with the previous behavior.

Comment (westonruter) [reply includes/templates/class-amp-post-template.php:346]: Workaround plugin to restore the previous behavior: https://gist.github.com/westonruter/559c7a9cbb9460d1099c3d5d9bfa5ad1

Comment (westonruter) [reply includes/templates/class-amp-post-template.php:346]: Fix: https://github.com/ampproject/amp-wp/pull/4547
